### ISSUE TITLE
add options for "import-jdl" subcommand to force/disable overwrites by path(s)

### ISSUE THREAD

**Issue by 62mkv** [jhipster/generator-jhipster#12054]
##### **Overview of the feature request**

This will allow to reduce human interaction for re-importing from JDL and reduce risk of errors 

##### **Motivation for or Use Case**

Every time "import-jdl" command is run, entities are "updated" even though nothing has changed in the JDL file about those entities definition, so it gets quite tedious to manually allow/disallow overwrites for every individual file. 

For example: if I run the `import-jdl` subcommand just to (re)create liquibase migrations, I could run it like this: 

`jhipster import-jdl application.jdl --allow-overwrite src/main/resources/config/liquibase --disallow-overwite *`

which basically says: whatever requires human interaction to allow overwrite, allow overwrite if it's in the `src/main/resources/config/liquibase` folder, otherwise disallow overwrite

##### **Related issues or PR**

see also #11896 

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by pascalgrimaud** [comment]
You could easily accept all changes, then use your own shell script to revert the folders you want

**Comment by mshima** [comment]
Closing following discussion at https://github.com/jhipster/generator-jhipster/issues/12080

**Comment by mshima** [comment]
This feature should be implemented upstream.

**Comment by 62mkv** [comment]
is corresponding issue already raised against yeoman-environment? can we link it here? 

also, I have a strong feeling, that there's much bigger issue, namely - "import-jdl" subcommand keeps updating entities that were not touched/changed since latest run. is there any way around it ? 

May be I could use some phased approach, where initially I updated just the `.jhipster/*.json` or `.yo-rc.json` definitions, and then generated code only for updated entities. I have seen that someone mentioned that 'import-jdl' is an alias, but an alias for what exactly? Also, @deepu105 (I believe) wrote somewhere about their vision of JDL-centric JHipster (which is de-facto how we use it in our team). But thus far it does not exactly look like that. Maybe v7 will change anything in this respect?

in other words: when I add an entity definition to a JDL file, I need to be able to ONLY generate code for this entity + relevant pieces of global code (like, menu, localization files, etc), but NOT any of the entity-specific code for entities that have already been processed by "import-jdl" 

is there any way of achieving this?

**Comment by mshima** [comment]
> is corresponding issue already raised against yeoman-environment?

No. I don't see that as a priority, so if your are planning to work on this raise it.
 
> also, I have a strong feeling, that there's much bigger issue, namely - "import-jdl" subcommand keeps updating entities that were not touched/changed since latest run. is there any way around it ?

I don't see this as a bug.

> May be I could use some phased approach, where initially I updated just the `.jhipster/*.json` or `.yo-rc.json` definitions, and then generated code only for updated entities.

Not doable with our current man power.
Need the inverse, we need jhipster-core to stop writing the file to the disk, so we could compare the old state.

> I have seen that someone mentioned that 'import-jdl' is an alias, but an alias for what exactly?

`jhipster jdl`

> Also, @deepu105 (I believe) wrote somewhere about their vision of JDL-centric JHipster (which is de-facto how we use it in our team). But thus far it does not exactly look like that. Maybe v7 will change anything in this respect? in other words: when I add an entity definition to a JDL file, I need to be able to ONLY generate code for this entity + relevant pieces of global code (like, menu, localization files, etc), but NOT any of the entity-specific code for entities that have already been processed by "import-jdl"
>
> is there any way of achieving this?

Not currently.
And I think the main goal for JDL-centric is to remove `.yo-rc.json` and `.jhipster/*.json`.

I think the easiest way to accomplish what you want is the .rc-resolver as discussed at the other topic.

**Comment by 62mkv** [comment]
concerning your comment on the "not a bug": this is what the official documentation has to say on this: 

> By default import-jdl regenerates only entities which have changed, if you want all your entities to be regenerated then pass in the --force flag. Please note that this will overwrite all your local changes to the entity files

(from https://www.jhipster.tech/creating-an-entity/) 

am I reading it wrong?

**Comment by 62mkv** [comment]
(also, `jhipster jdl` is an unknown command)

**Comment by 62mkv** [comment]
I've discovered a `-d` option, and it seems that `import-jdl` subcommand is indeed looking into only one (changed) entity, but it then proceeds to generating code (even though a `--json-only` has been given as an argument), and _that step_ is regenerating everything (including code entity-specific code for unchanged entities)

should I raise a defect for this?

**Comment by 62mkv** [comment]
that is probably what you mean when you say "Need the inverse, we need jhipster-core to stop writing the file to the disk, so we could compare the old state." 

So it indeed, looks like this is exactly what is happening. 

If I remember correctly, `jhipster-core` is going to be part of this project since v7 release. So maybe it's worth creating an issue for that in this project? And first of all, should the documentation be updated?

**Comment by mshima** [comment]
> concerning your comment on the "not a bug": this is what the official documentation has to say on this:
> 
> > By default import-jdl regenerates only entities which have changed, if you want all your entities to be regenerated then pass in the --force flag. Please note that this will overwrite all your local changes to the entity files
> 
> (from https://www.jhipster.tech/creating-an-entity/)
> 
> am I reading it wrong?

It's a bug then.
You can open a new issue.

> (also, `jhipster jdl` is an unknown command)

Should be there for jhipster >= 6.10.0

> I've discovered a `-d` option, and it seems that `import-jdl` subcommand is indeed looking into only one (changed) entity, but it then proceeds to generating code (even though a `--json-only` has been given as an argument), and _that step_ is regenerating everything (including entity-specific code for unchanged entities)
> 
> should I raise a defect for this?

Have you tried `--ignore-application`?
